drivers/led/pca9533.c

```diff
@@ -0,0 +1,364 @@
+	struct pca9533_data *data = dev->data;
+	int ret;
+	uint8_t ch, duty, psc, cur;
+	uint32_t period, duty32;
+
+	if (led >= PCA9533_CHANNELS) {
+		LOG_ERR("Invalid LED index: %u", led);
+		return -EINVAL;
+	}
+
+	period = delay_on + delay_off;
+	if (period < BLINK_MIN_MS || period > BLINK_MAX_MS) {
+		LOG_ERR("Invalid blink period: %u ms (min: %u, max: %u)", period, BLINK_MIN_MS,
+			BLINK_MAX_MS);
+		return -ENOTSUP;
+	}
+
+	/* Calculate duty cycle with overflow protection */
+	duty32 = (delay_on * 256U) / period;
+	duty = CLAMP(duty32, 0, UINT8_MAX);
+	psc = ms_to_psc(period);
+	cur = data->led_engine[led];
+
+	/* If LED is sole user of its engine, update in place */
+	if (cur < PCA9533_ENGINES && data->engine_users[cur] == BIT(led)) {
+		/* Only update if parameters changed */
+		if (data->pwm_val[cur] != duty || data->psc_val[cur] != psc) {
+			ret = i2c_reg_write_byte_dt(&config->i2c, cur ? PCA9533_PSC1 : PCA9533_PSC0,
+						    psc);
+			if (ret == 0) {
+				ret = i2c_reg_write_byte_dt(
+					&config->i2c, cur ? PCA9533_PWM1 : PCA9533_PWM0, duty);
+			}
+			if (ret == 0) {
+				data->psc_val[cur] = psc;
+				data->pwm_val[cur] = duty;
+			}
+		}
+		return 0;
+	}
+
+	/* Acquire new engine with desired parameters */
+	ret = engine_acquire(data, duty, psc, &ch);
+	if (ret) {
+		LOG_WRN("No PWM engine available for LED %u blink", led);
+		return ret;
+	}
+
+	/* If engine is new (no users), program it */
+	if (data->engine_users[ch] == 0) {
+		ret = i2c_reg_write_byte_dt(&config->i2c, ch ? PCA9533_PSC1 : PCA9533_PSC0, psc);
+		if (ret == 0) {
+			ret = i2c_reg_write_byte_dt(&config->i2c, ch ? PCA9533_PWM1 : PCA9533_PWM0,
+						    duty);
+		}
+		if (ret) {
+			LOG_ERR("Failed to program engine %u: %d", ch, ret);
+			return ret;
+		}
+		data->psc_val[ch] = psc;
+		data->pwm_val[ch] = duty;
+	}
+
+	LOG_DBG("LED%u now on engine %u (psc %u duty %u)", led, ch, psc, duty);
+	engine_release(data, led);
+	engine_bind(data, led, ch);
+	return ls_update(&config->i2c, led, ch ? LS_FUNC_PWM1 : LS_FUNC_PWM0);
+}
+
+static int pca9533_led_init_chip(const struct device *dev)
+{
+	struct pca9533_data *data = dev->data;
+
+	for (uint8_t i = 0; i < PCA9533_CHANNELS; i++) {
+		data->led_engine[i] = 0xFF;
+	}
+	data->engine_users[0] = 0;
+	data->engine_users[1] = 0;
+
+	/* The Power-On Reset already initializes the registers to their default state
+	 * no need to write them here. We'll just reset bookkeeping
+	 */
+
+	return 0;
+}
+
+static int pca9533_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+		return pca9533_led_init_chip(dev);
+	case PM_DEVICE_ACTION_RESUME:
+	case PM_DEVICE_ACTION_SUSPEND:
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return 0;
+
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int pca9533_led_init(const struct device *dev)
+{
+	const struct pca9533_config *config = dev->config;
+
+	if (!device_is_ready(config->i2c.bus)) {
```

Comment:
sorry just stopped one more, you can use `if (!i2c_is_ready_dt(&config->i2c)) {` here